scripts/count_decls.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2024 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Damiano Testa
+-/
+import Mathlib
+
+/-!
+# `count_decls` -- a tally of declarations in `Mathlib`
+
+This file is used by the `Periodic reports` script to generate a tally of declarations
+in `Mathlib`, `Batteries` and core.
+
+Source: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Metamathematics.3A.20Theorems.20by.20Domain.20Areas/near/374090639
+-/
+
+namespace PeriodicReports
+open Lean Elab Meta
+
+/-- A structure to tally declaration types:
+* `thms` are theorems;
+* `preds` are predicates;
+* `types` are types;
+* `data` is everything else.
+-/
+structure Tally where
+  thms  : HashSet Name := {}
+  data  : HashSet Name := {}
+  preds : HashSet Name := {}
+  types : HashSet Name := {}
+
+/-- `toString t` produces a string where each line is either
+* the (human-readable) name of a field of `t : Tally`, or
+* the name of a single declaration followed by `,`.
+
+The script to extract comparison data relies on the format
+* lines ending with `,` or not,
+* single declaration per line.
+-/
+def toString (t : Tally) : String :=
+  let print (h : HashSet Name) : String :=
+    let tot := h.toList.map (·.toString)
+    String.intercalate ",\n" tot ++ ","
+  s!"\
+Theorems
```

Comment:
Small warning: this string literal is relying on the current line endings for the file. For the current version of Lean, it'd be more reliable to have it be `Theorems\n\`, etc. However, in a future version all line endings will be normalized to `\n` so it won't matter.

You can decide whether this is a big deal or not. In its current form, it'll only affect Windows users who run this script who also have Git configured to normalize line endings to CRLF.

Reply:
I expect the script to only run in CI, but it is easy to use `\n`, so I will update!